Modules/posixmodule.c

```diff
@@ -7657,24 +7658,36 @@ check_CreateSymbolicLink(void)
     return Py_CreateSymbolicLinkW != NULL;
 }
 
+static int _PPathCchRemoveFileSpec_Initialized = 0;
+typedef HRESULT(__stdcall *PPathCchRemoveFileSpec)(PWSTR pszPath, size_t cchPath);
+static PPathCchRemoveFileSpec _PPathCchRemoveFileSpec;
+
 /* Remove the last portion of the path - return 0 on success */
-static int
+static void
 _dirnameW(WCHAR *path)
 {
     WCHAR *ptr;
-    size_t length = wcsnlen_s(path, MAX_PATH);
-    if (length == MAX_PATH) {
-        return -1;
+    size_t length = wcslen(path);
+
+    if (!_PPathCchRemoveFileSpec_Initialized) {
+        HMODULE pathapi = LoadLibraryW(L"api-ms-win-core-path-l1-1-0.dll");
```

Comment:
`LoadLibrary` calls should really be outside the GIL, as they can trigger a lot of work.

If we require that the GIL is held on entry to this function, then wrapping the entire contents of this initialization block in `Py_BEGIN_ALLOW_THREADS` should be sufficient. There's likely more lost that gained if we go too fine-grained with it.

Reply:
Ok, cool - I'll make that change and see what happens.